Unconditionally leak autograd threads.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53018 Unconditionally leak autograd threads.**

Fixes #48888, maybe.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26725185](https://our.internmc.facebook.com/intern/diff/D26725185)

ezyang committed:
Unconditionally leak autograd threads.

Fixes #48888, maybe.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented:
Do we have somewhere the reason why this logic was added in the first place?
This question was never answered on the PR that adds it https://github.com/pytorch/pytorch/pull/21438#discussion_r299044331 and there is no comment there explaining why we need this.

ezyang commented:
@albanD It's a hygiene thing. In an ideal world, you properly cleanup everything after yourself on shutdown. This way, if you e.g., load PyTorch dynamic library and then unload it, you end up with a net zero state; if you're leaking threads, this isn't true.

As you might imagine, this is a pretty niche use case. So since it's causing problems it's more trouble than it's worth. Looks like need to disable the relevant tests.

ezyang committed:
Update on "Unconditionally leak autograd threads."


Fixes #48888, maybe.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26725185](https://our.internmc.facebook.com/intern/diff/D26725185)

[ghstack-poisoned]

ezyang committed:
Update on "Unconditionally leak autograd threads."


Fixes #48888, maybe.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26725185](https://our.internmc.facebook.com/intern/diff/D26725185)

[ghstack-poisoned]

Balandat reviewed (approved):
This indeed fixes an issue for me with test binaries not properly shutting down after a successful test run. I can't really speak to the larger blast radius of this change though.

malfet commented:
@albanD I remember I had to fix logic in the shutdown sequence because it lead to occasional crashes (especially while running C++ tests)
Perhaps a reasonable solution would be to introduce an environment variable controllable timeout during shutdown (setting it say 10 sec sounds reasonable by default)

ezyang commented:
@malfet Whatever logic that was, this PR is bypassing it entirely because we no longer send the shutdown tasks. Are you suggesting we should avoid landing this task (so that we keep exercising the shutdown logic), or that someone should look into doing this later?

malfet commented:
@ezyang Can you please hold of landing this one and I will send PR with the timeout logic I have in mind ( that will still leak threads, but after the delay)

malfet commented:
How about something along the following lines: https://github.com/pytorch/pytorch/pull/53170
(With that PR invoking code with TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT=0 would be equivalent to current PR, but by default shutdown would be performed correctly)
